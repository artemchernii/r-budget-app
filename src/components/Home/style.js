import styled, { createGlobalStyle } from 'styled-components';

export const GlobalStyle = createGlobalStyle`
    * {
        box-sizing: border-box;
    }
    body {
        margin: 0;
        padding: 0;
    }
`;
export const Home = styled.div`
    width: 90%;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
`;

Home.displayName = 'Home';
