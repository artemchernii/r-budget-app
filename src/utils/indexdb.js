const DB_VERSION = 1;
const DB_NAME = 'budget';

indexedDB.db = null;

indexedDB.onerror = function (e) {
    console.log(e);
};

function open() {
    return new Promise(function (resolve, reject) {
        const request = indexedDB.open(DB_NAME, DB_VERSION);

        // We can only create Object stores in a versionchange transaction.
        request.onupgradeneeded = function (e) {
            const db = e.target.result;

            // A versionchange transaction is started automatically.
            e.target.transaction.onerror = indexedDB.onerror;

            if (db.objectStoreNames.contains(DB_NAME)) {
                db.deleteObjectStore(DB_NAME);
            }

            db.createObjectStore(DB_NAME, { keyPath: 'id' });
        };

        request.onsuccess = function (e) {
            indexedDB.db = e.target.result;
            resolve();
        };

        request.onerror = (e) => {
            reject(Error(e));
        };
    });
}

function addItem(item) {
    const db = indexedDB.db;
    const trans = db.transaction([DB_NAME], 'readwrite');
    const store = trans.objectStore(DB_NAME);

    const request = store.put(item);

    request.onsuccess = function (e) {
        console.log('success', { e, item });
    };

    request.onerror = function (e) {
        console.log('Error Adding: ', e);
    };
}

function deleteItem(id) {
    const request = indexedDB.db
        .transaction([DB_NAME], 'readwrite')
        .objectStore(DB_NAME)
        .delete(id);

    request.onsuccess = (event) => {
        console.log('its done', event);
    };
}

function getItems() {
    return new Promise((resolve, reject) => {
        const db = indexedDB.db;
        if (!db) {
            reject(Error('No db'));
        }
        const trans = db.transaction([DB_NAME], 'readwrite');
        const store = trans.objectStore(DB_NAME);

        const getAllRequest = store.getAll();

        getAllRequest.onsuccess = function () {
            resolve(getAllRequest.result);
        };
    });
}

function getData(start, total) {
    return new Promise(function (resolve, reject) {
        const db = indexedDB.db;
        const t = db.transaction([DB_NAME], 'readonly');
        const store = t.objectStore(DB_NAME);
        const transactions = [];
        let hasSkipped = false;

        store.openCursor(null, 'prev').onsuccess = function (e) {
            var cursor = e.target.result;
            if (!hasSkipped && start > 0) {
                hasSkipped = true;
                cursor.advance(start);
                return;
            }
            if (cursor) {
                transactions.push(cursor.value);
                if (transactions.length < total) {
                    cursor.continue();
                } else {
                    resolve(transactions);
                }
            } else {
                resolve(transactions);
            }
        };
        store.openCursor(null, 'prev').onerror = function (e) {
            reject(e);
        };
    });
}

function setFavor(id) {
    const store = indexedDB.db
        .transaction([DB_NAME], 'readwrite')
        .objectStore(DB_NAME);
    const request = store.get(id);
    request.onerror = (event) => {
        console.error('WTF', event.message);
    };

    request.onsuccess = (event) => {
        const data = event.target.result;
        store.put({
            ...data,
            isFavoured: !data.isFavoured,
        });
    };
}

export { open, addItem, getItems, deleteItem, setFavor, getData };
