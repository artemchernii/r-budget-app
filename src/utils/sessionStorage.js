export const saveToStorage = (name, data) => {
    if (!window || !localStorage) {
        return;
    }

    localStorage.setItem(name, JSON.stringify(data));
};

export const getFromStorage = (name) => {
    if (!window || !localStorage) {
        return;
    }

    try {
        return JSON.parse(localStorage.getItem(name));
    } catch (error) {
        console.error(error);
    }
};
