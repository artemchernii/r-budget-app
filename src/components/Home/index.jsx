import { useEffect, useState } from 'react';
import Balance from '../Balance';
import Transactions from '../Transactions/Transactions';
import Form from '../Form';
import { HomeDiv } from './style';
import ErrorBoundary from '../ErrorBoundaries';
import { addItem, getItems } from '../../utils/indexdb';

export default function Home() {
    const [balance, setBalance] = useState(0);
    const [transactions, setTransactions] = useState([]);
    useEffect(() => {
        getItems()
            .then((items) => {
                setBalance(
                    items.reduce((acc, item) => {
                        return Number(acc) + +item.value;
                    }, 0)
                );
                setTransactions(items);
            })
            .catch((e) => {
                console.log('error here', e);
                return console.error(e);
            });
    }, [setTransactions]);

    const handleSubmit = ({ balance, date, comment }) => {
        const transaction = {
            label: 'Balance increased by',
            value: +balance,
            date,
            comment,
            id: Date.now(),
        };

        setBalance((c) => c + +balance);
        setTransactions((currentTransactions) => [
            transaction,
            ...currentTransactions,
        ]);
        addItem(transaction);
    };
    return (
        <ErrorBoundary>
            <HomeDiv>
                <Balance data-testid="balance" balance={+balance} />
                <Form handleSubmit={handleSubmit} />
                {transactions.length > 0 && (
                    <Transactions transactions={transactions} />
                )}
            </HomeDiv>
        </ErrorBoundary>
    );
}
