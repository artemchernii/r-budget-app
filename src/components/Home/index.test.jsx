import { beforeEach, describe, expect, it, vi } from 'vitest';
import Home from './index';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { getItems } from '../../utils/indexdb';

vi.mock('../../utils/indexdb', () => ({
    addItem: vi.fn(),
    getItems: vi
        .fn()
        .mockImplementation(() =>
            Promise.resolve([{ value: 100, id: '123123EEWQEQWEQ123123' }])
        ),
}));

describe('Home component', () => {
    let props;
    let sut;
    let container;

    beforeEach(async () => {
        await act(async () => {
            sut = render(<Home {...props} />);
        });
        container = sut.container;
    });

    describe('when transactions are returned successfully', () => {
        it('should render transactions', () => {
            expect(screen.getByTitle('transactions')).toBeInTheDocument();
        });
        it('should render transaction with one item with value got from db', async () => {
            await waitFor(() => {
                expect(screen.getByTestId('Value').textContent).toBe('100.00');
            });
        });
    });

    describe('when transaction is added', () => {
        it('should increase the balance', () => {
            const submit = sut.getByText('Save');
            const balanceInput = container.querySelector('[name="balance"]');
            const balanceContainer = sut.getByTestId('balance');
            fireEvent.change(balanceInput, { target: { value: 100 } });
            fireEvent.submit(submit);
            expect(balanceContainer.textContent).toBe('Current balance: 200');
        });
    });

    describe('when transactions are returned with error', () => {
        let consoleSpy;
        beforeEach(() => {
            consoleSpy = vi.spyOn(console, 'error');
            getItems.mockImplementation(() => Promise.reject('WTF'));
        });
        beforeEach(async () => {
            await act(async () => {
                sut = render(<Home {...props} />);
            });
            container = sut.container;
        });
        it('should throw an error', () => {
            expect(consoleSpy).toHaveBeenCalledWith('WTF');
        });
        it('should return transactions with zero items', () => {
            expect(screen.getByTitle('transactions')).toBeUndefined();
        });
    });
});
