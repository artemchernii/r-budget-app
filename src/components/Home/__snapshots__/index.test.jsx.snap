// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home component > when transactions are returned successfully > should render transaction with one item 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-gKAaef ddddNx"
      >
        <div>
          <div
            class="sc-bdnyFh dQVlVI"
            data-testid="balance"
            data-title="balance"
            title="balance"
          >
            Current balance: 
            0
          </div>
        </div>
        <form
          title="test-form"
        >
          <div
            class="input-container"
          >
            <input
              class="material-input"
              name="date"
              required=""
              type="date"
              value="2023-11-08"
            />
            <label
              class="material-label"
            >
              Date
            </label>
          </div>
          <div
            class="input-container"
          >
            <input
              class="material-input"
              id="balance"
              name="balance"
              required=""
              type="number"
              value="0"
            />
            <label
              class="material-label"
              for="balance"
            >
              New sum
            </label>
          </div>
          <div
            class="input-container"
          >
            <textarea
              class="material-input"
              name="comment"
              required=""
            />
            <label
              class="material-label"
            >
              Comment
            </label>
          </div>
          <button
            type="submit"
          >
            Save
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-gKAaef ddddNx"
    >
      <div>
        <div
          class="sc-bdnyFh dQVlVI"
          data-testid="balance"
          data-title="balance"
          title="balance"
        >
          Current balance: 
          0
        </div>
      </div>
      <form
        title="test-form"
      >
        <div
          class="input-container"
        >
          <input
            class="material-input"
            name="date"
            required=""
            type="date"
            value="2023-11-08"
          />
          <label
            class="material-label"
          >
            Date
          </label>
        </div>
        <div
          class="input-container"
        >
          <input
            class="material-input"
            id="balance"
            name="balance"
            required=""
            type="number"
            value="0"
          />
          <label
            class="material-label"
            for="balance"
          >
            New sum
          </label>
        </div>
        <div
          class="input-container"
        >
          <textarea
            class="material-input"
            name="comment"
            required=""
          />
          <label
            class="material-label"
          >
            Comment
          </label>
        </div>
        <button
          type="submit"
        >
          Save
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
